### ISSUE TITLE
Streaming audio from a slow internet connection

### ISSUE THREAD

**Issue by natloh** [Jrawke/mumbleExpress#31]
If a user has an slow internet connection their audio does not come through. On the receiving end with a fast connection you hear choppy bits of sound. On the receiving end on the slow connection they hear the audio from the fast connection clearly.

**Comment by Jrawke** [comment]
I think this is expected? If a user has an upload speed lower than the bandwidth needed for audio transmission then I think we should drop the latent buffers

However, maybe the buffer sizes in commit 9f1b0749aec846db826577ee4bb3459ad0263d27 need some tuning

What do you think?

**Comment by Jrawke** [comment]
Maybe this is what we should be using https://webrtc.org/

Needs some research

**Comment by tomKPZ** [comment]
On 01/30/2017 01:38 AM, Jrawke wrote:
>
> Maybe this is what we should be using https://webrtc.org/
>
Not sure webrtc would help in this case.  It's good for interclient 
communication, but it may not work well for us since we still need to 
send/receive data from the mumble server which AFAIK does not have 
webrtc support.

**Comment by tomKPZ** [comment]
On 01/30/2017 01:33 AM, Jrawke wrote:
>
> I think this is expected? If a user has an upload speed lower than the 
> bandwidth needed for audio transmission then I think we should drop 
> the latent buffers
>
Yeah, not sure there's much we can do about this :(
If we detect a slow upload speed, one thing we could try is lowering the 
bitrate client-side, and convert the bitrate using our server to the 
bitrate mumble supports.

We still need that bitrate converter too.  I haven't done my research so 
I don't know how complex it will be.  Probably there's a library for this?

**Comment by Jrawke** [comment]
Yeah, if there is no alternative to streaming audio over websockets the solution is becoming more clear now.

We will need to compress the audio client side, and add support on the backend to forward the encoded buffers. There will be libraries to help with this
